Mathlib/LinearAlgebra/Vandermonde.lean

```diff
@@ -0,0 +1,184 @@
+import Mathlib.LinearAlgebra.Matrix.Nondegenerate
+
+/-!
+# Vandermonde matrix
+
+This file defines the `vandermonde` matrix and gives its determinant.
+
+## Main definitions
+
+ - `vandermonde v`: a square matrix with the `i, j`th entry equal to `v i ^ j`.
+
+## Main results
+
+ - `det_vandermonde`: `det (vandermonde v)` is the product of `v i - v j`, where
+   `(i, j)` ranges over the unordered pairs.
+-/
+
+
+variable {R : Type _} [CommRing R]
+
+open Equiv Finset
+
+open BigOperators Matrix
+
+namespace Matrix
+
+/-- `vandermonde v` is the square matrix with `i`th row equal to `1, v i, v i ^ 2, v i ^ 3, ...`.
+-/
+def vandermonde {n : ℕ} (v : Fin n → R) : Matrix (Fin n) (Fin n) R := fun i j => v i ^ (j : ℕ)
+#align matrix.vandermonde Matrix.vandermonde
+
+@[simp]
+theorem vandermonde_apply {n : ℕ} (v : Fin n → R) (i j) : vandermonde v i j = v i ^ (j : ℕ) :=
+  rfl
+#align matrix.vandermonde_apply Matrix.vandermonde_apply
+
+@[simp]
+theorem vandermonde_cons {n : ℕ} (v0 : R) (v : Fin n → R) :
+    vandermonde (Fin.cons v0 v : Fin n.succ → R) =
+      Fin.cons (fun (j : Fin n.succ) => v0 ^ (j : ℕ)) fun i => Fin.cons 1
+      fun j => v i * vandermonde v i j := by
+  ext (i j)
+  refine' Fin.cases (by simp) (fun i => _) i
+  refine' Fin.cases (by simp) (fun j => _) j
+  simp [pow_succ]
+#align matrix.vandermonde_cons Matrix.vandermonde_cons
+
+theorem vandermonde_succ {n : ℕ} (v : Fin n.succ → R) :
+    vandermonde v =
+      Fin.cons (fun (j : Fin n.succ) => v 0 ^ (j : ℕ)) fun i =>
+        Fin.cons 1 fun j => v i.succ * vandermonde (Fin.tail v) i j := by
+  conv_lhs => rw [← Fin.cons_self_tail v, vandermonde_cons]
+#align matrix.vandermonde_succ Matrix.vandermonde_succ
+
+theorem vandermonde_mul_vandermonde_transpose {n : ℕ} (v w : Fin n → R) (i j) :
+    (vandermonde v ⬝ (vandermonde w)ᵀ) i j = ∑ k : Fin n, (v i * w j) ^ (k : ℕ) := by
+  simp only [vandermonde_apply, Matrix.mul_apply, Matrix.transpose_apply, mul_pow]
+#align matrix.vandermonde_mul_vandermonde_transpose Matrix.vandermonde_mul_vandermonde_transpose
+
+theorem vandermonde_transpose_mul_vandermonde {n : ℕ} (v : Fin n → R) (i j) :
+    ((vandermonde v)ᵀ ⬝ vandermonde v) i j = ∑ k : Fin n, v k ^ (i + j : ℕ) := by
+  simp only [vandermonde_apply, Matrix.mul_apply, Matrix.transpose_apply, pow_add]
+#align matrix.vandermonde_transpose_mul_vandermonde Matrix.vandermonde_transpose_mul_vandermonde
+
+theorem det_vandermonde {n : ℕ} (v : Fin n → R) :
+    det (vandermonde v) = ∏ i : Fin n, Finset.prod (Ioi i) (fun j => v j - v i) := by
+  unfold vandermonde
+  induction' n with n ih
+  · exact det_eq_one_of_card_eq_zero (Fintype.card_fin 0)
+  calc
+    det (of fun i j : Fin n.succ => v i ^ (j : ℕ)) =
+        det
+          (of fun i j : Fin n.succ =>
+            Matrix.vecCons (v 0 ^ (j : ℕ)) (fun i => v (Fin.succ i) ^ (j : ℕ) - v 0 ^ (j : ℕ)) i) :=
+      by
+      refine' det_eq_of_forall_row_eq_smul_add_const (Matrix.vecCons 0 1) 0 (Fin.cons_zero _ _) _
```

Comment:
At very least we should have a porting note here noting that this is much more complicated than the mathlib3 proof.

Possibly we could factor it out as a lemma? This calc block is getting out of hand.

Reply:
Oh, I see! In the mathlib3 proof, the calc block left two goals for later (the curly brace blocks after the `calc`). You can do this in mathlib4 as well, but you need to use `?_` instead for the goals you want to postpone.

Reply:
I did not know I could postpone a goal in mathlib4 as well. Thank you!